src/test/scala/firrtlTests/stage/FirrtlStageSpec.scala

```diff
@@ -0,0 +1,415 @@
+  info("I want to receive error messages when I do not specify mandatory inputs")
+  feature("FirrtlStage input validation of mandatory options") {
+    scenario("User gives no command line options (no input circuit specified)") {
+      val f = new FirrtlStageFixture
+
+      When("the user passes no arguments")
+      val (out, err, result) = grabStdOutErr { catchStatus { f.stage.main(Array.empty) } }
+
+      Then("an error should be printed on stdout")
+      out should include (s"Error: Unable to determine FIRRTL source to read")
+
+      And("no usage text should be shown")
+      out should not include ("Usage: firrtl")
+
+      And("nothing should print to stderr")
+      err should be (empty)
+
+      And("the exit code should be 1")
+      result should be (Left(1))
+    }
+  }
+
+  info("As a FIRRTL command line user")
+  info("I want to receive helpful error and warnings message")
+  feature("FirrtlStage input validation") {
+    /* Note: most input validation occurs inside firrtl.stage.phases.Checks. This seeks to validate command line
+     * behavior.
+     */
+
+    scenario("User tries to use an implicit annotation file") {
+      val f = new FirrtlStageFixture
+      val td = new TargetDirectoryFixture("implict-annotation-file")
+      val circuit = new SimpleFirrtlCircuitFixture
+
+      And("implicit legacy and extant annotation files")
+      val annoFiles = Array( (new File(td.dir + "/Top.anno"), "/annotations/SampleAnnotations.anno"),
+                             (new File(td.dir + "/Top.anno.json"), "/annotations/SampleAnnotations.anno.json") )
+      annoFiles.foreach{ case (file, source) => copyResourceToFile(source, file) }
+
+      When("the user implies an annotation file (an annotation file has the same base name as an input file)")
+      val in = new File(td.dir + "/Top.fir")
+      val pw = new PrintWriter(in)
+      pw.write(circuit.input)
+      pw.close()
+      val (out, _, result) = grabStdOutErr{ catchStatus { f.stage.main(Array("-td", td.dir.toString,
+                                                                             "-i", in.toString,
+                                                                             "-foaf", "Top.out",
+                                                                             "-X", "high",
+                                                                             "-E", "high")) } }
+
+      Then("the implicit annotation file should NOT be read")
+      val annoFileOut = new File(td.dir + "/Top.out.anno.json")
+      val annotationJson = Source.fromFile(annoFileOut).mkString
+      annotationJson should not include ("InlineInstances")
+
+      And("no warning should be printed")
+      out should not include ("Warning:")
+
+      And("no error should be printed")
+      out should not include ("Error:")
+
+      And("the exit code should be 0")
+      result shouldBe a [Right[_,_]]
+    }
+
+    scenario("User provides unsupported legacy annotations") {
```

Comment:
This scenario thing is really cool, I would be good to hear about your experience with it.

Reply:
All the scaladoc testing styles are interesting. The use of `FeatureSpec with `GivenWhenThen` is that it enables more "acceptance" style testing. You can get somewhat creative with trying to think of how a user would use a `Stage` from the command line versus testing all the individual components. I generally think this makes sense for testing user-exposed `Stage`s, but not so much for the internal stuff. I usually like `FlatSpec` with a lot of `info()` methods there.

Generally, it may be worth a discussion at some point about aligning to some testing style.